### ISSUE TITLE
`constraint-generators` test failure on CI: `prop_univSound @BaseFn`

### ISSUE THREAD

**Issue by lehins** [IntersectMBO/cardano-ledger#4721]
Recently we started to get test failures on CI for a test case that had no problems in months. It sounds like some recent change introduced a significant overhead leading to long runtimes. Here is one reproducer, but there are many more failures on CI. This started to happen quite often and needs to be fixed. We've tried doubling the time in #4718, which didn't help.
```
  1) constrained prop_univSound @BaseFn
       Timeout of 20000000 microseconds exceeded. (after 605 tests and 1 shrink):
         Rng :: Map Int Int -> [Int]
         
         fn ctx = Rng v_0
         
         spec = TypeSpec
                  (ListSpec
                     hint = Nothing
                     must = []
                     size = MemberSpec [ 0 elements ...] @Integer
                     elem = constrained $ \ v_1 ->
                              exists v_0 in
                              {whenTrue v_0 $ assert ["toPred Term"] $ LessOrEqual (-3) v_1
                               whenTrue (Not v_0) $ assert ["toPred Term"] $ LessOrEqual (-3) v_1}
                     fold = FoldSpec
                              fn   = Id
                              spec = constrained $ \ v_1 ->
                                       exists v_0 in
                                       {whenTrue v_0 $
                                          {assert ["toPred Term"] $ LessOrEqual 0 v_1
                                           assert ["toPred Term"] $ LessOrEqual v_1 4}
                                        whenTrue (Not v_0) $
                                          {assert ["toPred Term"] $ LessOrEqual 0 v_1
                                           assert ["toPred Term"] $ LessOrEqual v_1 4}})
                  []
         
         propagateSpecFun fn ctx spec =
           TypeSpec
             (MapSpec
                hint       = Nothing
                mustKeys   = fromList []
                mustValues = []
                size       = MemberSpec [ 0 elements ...] @Integer
                elem       = constrained $ \ v_2 ->
                               [Snd v_2/v_1]exists v_0 in
                                            {whenTrue v_0 $
                                               assert ["toPred Term"] $ LessOrEqual (-3) v_1
                                             whenTrue (Not v_0) $
                                               assert ["toPred Term"] $ LessOrEqual (-3) v_1}
                fold       = FoldSpec
                               fn   = Id
                               spec = constrained $ \ v_1 ->
                                        exists v_0 in
                                        {whenTrue v_0 $
                                           {assert ["toPred Term"] $ LessOrEqual 0 v_1
                                            assert ["toPred Term"] $ LessOrEqual v_1 4}
                                         whenTrue (Not v_0) $
                                           {assert ["toPred Term"] $ LessOrEqual 0 v_1
                                            assert ["toPred Term"] $ LessOrEqual v_1 4}})
             []
         
         Simplfied spec:
           TypeSpec
             (MapSpec
                hint       = Nothing
                mustKeys   = fromList []
                mustValues = []
                size       = MemberSpec [ 0 elements ...] @Integer
                elem       = constrained $ \ v_2 ->
                               [Snd v_2/v_1]exists v_0 in
                                            {whenTrue v_0 $
                                               assert ["toPred Term"] $ LessOrEqual (-3) v_1
                                             whenTrue (Not v_0) $
                                               assert ["toPred Term"] $ LessOrEqual (-3) v_1}
                fold       = FoldSpec
                               fn   = Id
                               spec = constrained $ \ v_1 ->
                                        exists v_0 in
                                        {whenTrue v_0 $
                                           {assert ["toPred Term"] $ LessOrEqual 0 v_1
                                            assert ["toPred Term"] $ LessOrEqual v_1 4}
                                         whenTrue (Not v_0) $
                                           {assert ["toPred Term"] $ LessOrEqual 0 v_1
                                            assert ["toPred Term"] $ LessOrEqual v_1 4}})
             []

  To rerun use: --match "/constrained/prop_univSound @BaseFn/" --seed 1522016486
```

Other recent reproducers:
* https://github.com/IntersectMBO/cardano-ledger/actions/runs/11508716952/job/32039500171

**Comment by neilmayhew** [comment]
I did some git bisecting and it seems that the problem was introduced (or made bad enough to fail often enough to notice) in d24211a6d8a04222a7b9add7a1d9157f2d631933

**Comment by neilmayhew** [comment]
Note that if you have a fast machine, a given seed that fails in CI may not fail locally. My machine is almost 2x the speed of the CI runners, and I had to set the time limit back down to 10s to get it to fail with the seed shown in the description (`1522016486`).

**Comment by MaximilianAlgehed** [comment]
I have a fix for this in #4730. The issue was that the generator for lists with sum constraints was changed not to fail immediately when the size constraint is difficult to meet. The result was that we ended up running way too many checks on impossible to satisfy size constraints.
